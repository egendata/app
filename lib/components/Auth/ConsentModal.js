import React from 'react'
import { Image, View, TouchableOpacity } from 'react-native'
import Collapsible from 'react-native-collapsible'
import Modal from 'react-native-modal'
import { Wrap, ScrollViewWrap } from '../view/Wrapper'
import styled, { theme as globalTheme } from '../../theme'
import { H3, H4, Paragraph, Separator } from '../typography/Typography'
import {
  AcceptConsentButton,
  DenyConsentButton,
  ConsentButtonWrap,
} from '../elements/Button/ConsentButton'
import { Icon } from '../elements/Icon/Icon'

const StyledModal = styled(Modal)`
  margin: 0;
  width: 100%;
  height: 100%;
`

const ModalWrapper = styled(View)`
  margin-top: auto;
  height: 85%;
`

const ConsentHeader = styled(View)`
  background-color: ${({ theme }) => theme.colors.white};
  padding: 24px 36px;
  flex-direction: row;
  align-self: stretch;
`

const ClientDescription = styled(View)`
  padding: 4px 0;
`

/* TODO(@all): much of the commented out code in this component,
 * can probably be used for showing given/active consents */
/*
const ConsentHeaderExternal = styled(ConsentHeader)`
  padding: 12px 36px;
`
*/

const ScopeItemWrapper = styled(View)`
  margin-bottom: 0px;
`

const readAndWriteHelper = type => {
  switch (type) {
    case 'READ':
      return 'L채sa'
    case 'WRITE':
      return 'Skriva'
    default:
      return 'Inga r채ttigheter'
  }
}

const ScopeItem = ({ scope, lastItem }) => {
  const [collapsed, setCollapsed] = React.useState(true)
  const { read, write } = scope

  return (
    <ScopeItemWrapper key={scope.area} style={{ paddingHorizontal: 36 }}>
      <TouchableOpacity
        activeOpactiy={1}
        onPress={_ => setCollapsed(wasCollapsed => !wasCollapsed)}
      >
        <View
          style={{
            alignSelf: 'stretch',
            flexDirection: 'row',
            alignItems: 'center',
            justifyContent: 'space-between',
          }}
        >
          <Paragraph small>{scope.description}</Paragraph>
          <Icon width="14" height="14" name="info" />
        </View>
      </TouchableOpacity>
      <Collapsible collapsed={collapsed}>
        {read && (
          <>
            <H4 style={{ fontSize: 12, marginTop: 8 }}>
              {readAndWriteHelper(read.type)}
            </H4>
            <Paragraph small>{read.purpose}</Paragraph>
          </>
        )}

        {write && (
          <>
            <H4 style={{ fontSize: 12, marginTop: 8 }}>
              {readAndWriteHelper(write.type)}
            </H4>
            <Paragraph small>{write.description}</Paragraph>
          </>
        )}
      </Collapsible>
      {lastItem ? (
        <View style={{ marginTop: 12, marginBottom: 16 }} />
      ) : (
        <Separator style={{ marginBottom: 12, marginTop: 12 }} />
      )}
    </ScopeItemWrapper>
  )
}

const ConsentModal = ({
  data: { viewModel },
  visible,
  onApprove,
  onReject,
}) => {
  const handleApprovePermissions = () => onApprove({ local: viewModel.local })

  // const { local = [], external = [] } = data
  const { local = [] } = viewModel

  return (
    <Wrap>
      <StyledModal
        animationType="slide"
        isVisible={visible}
        backdropOpacity={0.6}
      >
        <ModalWrapper>
          <Separator style={{ marginBottom: 0, marginTop: 0 }} />
          <ConsentHeader>
            <Image
              source={{ uri: viewModel.iconURI }}
              style={{ width: 64, height: 64, marginRight: 12 }}
            />
            <ClientDescription>
              <H3 style={{ marginBottom: 8 }}>{viewModel.displayName}</H3>
              <Paragraph align="left">{viewModel.description}</Paragraph>
            </ClientDescription>
          </ConsentHeader>
          <Separator style={{ marginBottom: 0, marginTop: 0 }} />
          <ScrollViewWrap
            style={{
              paddingTop: 24,
              backgroundColor: globalTheme.colors.quiet,
            }}
            contentContainerStyle={{
              width: '100%',
              alignItems: 'stretch',
            }}
          >
            {/* <Paragraph style={{ paddingHorizontal: 32 }}> */}
            {/*   Vill uppr채tta en relation med dig */}
            {/* </Paragraph> */}

            {local.map((permission, i) => (
              <ScopeItem
                key={permission.area}
                lastItem={i === local.length - 1}
                scope={permission}
              />
            ))}

            {/* external.map(({ scope }) => (
              <View key={scope.displayName}>
                <Separator style={{ marginBottom: 0, marginTop: 0 }} />
                <ConsentHeaderExternal>
                  <H4>{scope.displayName}</H4>
                  <Paragraph align="left" small>
                    {scope.description}
                  </Paragraph>
                </ConsentHeaderExternal>
                <Separator style={{ marginBottom: 24, marginTop: 0 }} />
                {scope.areas.map((scope, i) => (
                  <ScopeItem key={scope.area} scope={scope} />
                ))}
              </View>
            )) */}
          </ScrollViewWrap>
          <ConsentButtonWrap>
            <AcceptConsentButton onPress={handleApprovePermissions}>
              Till책t
            </AcceptConsentButton>
            <Separator
              vertical={true}
              style={{ marginBottom: 0, marginTop: 0 }}
            />
            <DenyConsentButton onPress={onReject}>Neka</DenyConsentButton>
          </ConsentButtonWrap>
        </ModalWrapper>
      </StyledModal>
    </Wrap>
  )
}

export default ConsentModal
