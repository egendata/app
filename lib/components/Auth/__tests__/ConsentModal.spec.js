import React from 'react'
import { render } from '@testing-library/react-native'
import MockedProvider from '../../../utils/test-utils'
import ConsentModal from '../ConsentModal'

const mock = {
  displayName: 'My CV',
  description: 'An app for your CV online',
  iconURI: 'https://mycv.work/android-icon-96x96.png',
  local: [
    {
      area: 'baseData',
      description: 'Personal information.',
      read: {
        area: 'baseData',
        domain: 'https://mycv.work',
        id: '18710e28-7d6c-49cf-941e-0f954bb179ae',
        lawfulBasis: 'CONSENT',
        purpose: 'In order to create a CV using our website.',
        type: 'READ',
        kid:
          'http://localhost:4000/jwks/enc_1d2cbc575751efb6e498f5c00c5108d48fa94dbf3994fa9ea06bbf3c97f6aa05',
      },
      write: {
        area: 'baseData',
        description: 'Personal information.',
        domain: 'https://mycv.work',
        id: '1712ec0c-9ae6-472f-9e14-46088e51f505',
        lawfulBasis: 'CONSENT',
        type: 'WRITE',
      },
    },
    {
      area: 'experience',
      description: 'A list of your work experiences.',
      read: {
        area: 'experience',
        purpose: 'In order to create a CV using our website.',
        domain: 'https://mycv.work',
        id: '55c24372-6956-4891-b5ff-a6cf69fb5c8b',
        lawfulBasis: 'CONSENT',
        type: 'READ',
        kid:
          'http://localhost:4000/jwks/enc_1d2cbc575751efb6e498f5c00c5108d48fa94dbf3994fa9ea06bbf3c97f6aa05',
      },
      write: {
        area: 'experience',
        description: 'A list of your work experiences.',
        domain: 'https://mycv.work',
        id: 'd5dab30d-b0ac-43e3-9ac8-cff8b39ca560',
        lawfulBasis: 'CONSENT',
        type: 'WRITE',
      },
    },
  ],
  external: [
    {
      area: 'experience',
      read: {
        area: 'experience',
        purpose: 'In order to create a CV using our website.',
        domain: 'https://national.gov',
        id: 'fc284cf5-b1af-4fac-b793-7d1adf8a9c60',
        lawfulBasis: 'CONSENT',
        type: 'READ',
        kid:
          'http://localhost:4000/jwks/enc_1d2cbc575751efb6e498f5c00c5108d48fa94dbf3994fa9ea06bbf3c97f6aa05',
      },
    },
  ],
}

describe('components/Consent/ConsentModal', () => {
  it('should render without errors', () => {
    const { container } = render(
      <MockedProvider>
        <ConsentModal visible={false} data={{ viewModel: mock }} />
      </MockedProvider>,
    )

    expect(container).toMatchSnapshot()
  })
})
