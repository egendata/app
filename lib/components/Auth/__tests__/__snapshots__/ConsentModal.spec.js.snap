// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Consent/ConsentModal should render without errors 1`] = `
<ForwardRef>
  <ForwardRef
    animationType="slide"
    backdropOpacity={0.6}
    isVisible={true}
  >
    <ForwardRef>
      <ForwardRef
        style={
          Object {
            "marginBottom": 0,
            "marginTop": 0,
          }
        }
      />
      <ForwardRef>
        <Image
          source={
            Object {
              "uri": "https://mycv.work/android-icon-96x96.png",
            }
          }
          style={
            Object {
              "height": 64,
              "marginRight": 12,
              "width": 64,
            }
          }
        />
        <ForwardRef>
          <ForwardRef
            style={
              Object {
                "marginBottom": 8,
              }
            }
          >
            My CV
          </ForwardRef>
          <Paragraph
            align="left"
          >
            An app for your CV online
          </Paragraph>
        </ForwardRef>
      </ForwardRef>
      <ForwardRef
        style={
          Object {
            "marginBottom": 0,
            "marginTop": 0,
          }
        }
      />
      <ScrollViewWrap
        contentContainerStyle={
          Object {
            "alignItems": "stretch",
            "width": "100%",
          }
        }
        style={
          Object {
            "backgroundColor": "#FAFAFA",
            "paddingTop": 24,
          }
        }
      >
        <ScopeItem
          lastItem={false}
          scope={
            Object {
              "area": "baseData",
              "description": "Personal information.",
              "read": Object {
                "area": "baseData",
                "domain": "https://mycv.work",
                "id": "18710e28-7d6c-49cf-941e-0f954bb179ae",
                "kid": "http://localhost:4000/jwks/enc_1d2cbc575751efb6e498f5c00c5108d48fa94dbf3994fa9ea06bbf3c97f6aa05",
                "lawfulBasis": "CONSENT",
                "purpose": "In order to create a CV using our website.",
                "type": "READ",
              },
              "write": Object {
                "area": "baseData",
                "description": "Personal information.",
                "domain": "https://mycv.work",
                "id": "1712ec0c-9ae6-472f-9e14-46088e51f505",
                "lawfulBasis": "CONSENT",
                "type": "WRITE",
              },
            }
          }
        />
        <ScopeItem
          lastItem={true}
          scope={
            Object {
              "area": "experience",
              "description": "A list of your work experiences.",
              "read": Object {
                "area": "experience",
                "domain": "https://mycv.work",
                "id": "55c24372-6956-4891-b5ff-a6cf69fb5c8b",
                "kid": "http://localhost:4000/jwks/enc_1d2cbc575751efb6e498f5c00c5108d48fa94dbf3994fa9ea06bbf3c97f6aa05",
                "lawfulBasis": "CONSENT",
                "purpose": "In order to create a CV using our website.",
                "type": "READ",
              },
              "write": Object {
                "area": "experience",
                "description": "A list of your work experiences.",
                "domain": "https://mycv.work",
                "id": "d5dab30d-b0ac-43e3-9ac8-cff8b39ca560",
                "lawfulBasis": "CONSENT",
                "type": "WRITE",
              },
            }
          }
        />
      </ScrollViewWrap>
      <ForwardRef>
        <AcceptConsentButton
          onPress={[Function]}
        >
          Till√•t
        </AcceptConsentButton>
        <ForwardRef
          style={
            Object {
              "marginBottom": 0,
              "marginTop": 0,
            }
          }
          vertical={true}
        />
        <DenyConsentButton>
          Neka
        </DenyConsentButton>
      </ForwardRef>
    </ForwardRef>
  </ForwardRef>
</ForwardRef>
`;
